### PR TITLE
Support all bucketing types for hive temporary table

### PR DESCRIPTION
**Pull request by viczhang861**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java [hunk 1]
```diff
@@ -13,33 +13,65 @@
...
+     *  for json serialization only, don't use this method to create HivePartitioningHandle
+     */
     @JsonCreator
     public HivePartitioningHandle(
             @JsonProperty("bucketCount") int bucketCount,
-            @JsonProperty("hiveTypes") List<HiveType> hiveTypes,
-            @JsonProperty("maxCompatibleBucketCount") OptionalInt maxCompatibleBucketCount)
+            @JsonProperty("maxCompatibleBucketCount") OptionalInt maxCompatibleBucketCount,
+            @JsonProperty("hiveTypes") Optional<List<HiveType>> hiveTypes,
+            @JsonProperty("types") Optional<List<Type>> types)
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java:65]: Let's accept `BucketingType` as a parameter and check if `types` or `hiveTypes` are present based on the `BucketingType`.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java [hunk 2]
```diff
@@ -13,33 +13,65 @@
...
         implements ConnectorPartitioningHandle
 {
     private final int bucketCount;
-    private final List<HiveType> hiveTypes;
+    private final Optional<List<HiveType>> hiveTypes;
+    private final Optional<List<Type>> types;
     private final OptionalInt maxCompatibleBucketCount;
+    private final HiveBucketProperty.BucketingType bucketingType;
 
+    public static HivePartitioningHandle createPartitioningHandleWithHiveType(
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java:41]: How about `createHiveCompatiblePartitioningHandle`

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java [hunk 3]
```diff
@@ -13,33 +13,65 @@
...
+            int bucketCount,
+            List<HiveType> hiveTypes,
+            OptionalInt maxCompatibleBucketCount)
+    {
+        return new HivePartitioningHandle(bucketCount, maxCompatibleBucketCount, Optional.of(hiveTypes), Optional.empty());
+    }
+
+    public static HivePartitioningHandle createPartitioningHandleWithPrestoType(
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java:49]: `createPrestoNativePartitioningHandle`

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java [hunk 4]
```diff
@@ -49,21 +81,33 @@ public int getBucketCount()
...
+    }
+
     @JsonProperty
     public OptionalInt getMaxCompatibleBucketCount()
     {
         return maxCompatibleBucketCount;
     }
 
+    @JsonProperty
+    public HiveBucketProperty.BucketingType getBucketingType()
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java:102]: Let's extract `BucketingType` into a standalone class. Let's call it `BucketFunctionType`.

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java [hunk 5]
```diff
@@ -49,21 +81,33 @@ public int getBucketCount()
...
+    {
+        return bucketingType;
+    }
+
     @Override
     public String toString()
     {
-        return format("buckets=%s, hiveTypes=%s", bucketCount, hiveTypes);
+        return format("buckets=%s, hiveTypes=%s, types=%s, bucketingType=%s", bucketCount, hiveTypes, types, bucketingType);
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitioningHandle.java:110]: This is what is printed in the plan. In a current form it might be a little hard to read. How about we merge `hiveTypes` and `types` and print only one?

e.g.:

```
format("bucketFunctionType=%s, buckets=%s, types=%s", bucketingType, bucketCount, hiveTypes or types);
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 6]
```diff
@@ -2217,14 +2240,12 @@ public ConnectorTableLayoutHandle getAlternativeLayoutHandle(ConnectorSession se
...
-                        .map(type -> toHiveType(
-                                typeTranslator,
-                                translateHiveUnsupportedTypeForTemporaryTable(type, typeManager)))
-                        .collect(toImmutableList()),
-                OptionalInt.empty());
+        if (!partitionTypes.isEmpty()) {
+            return createPartitioningHandleWithPrestoType(partitionCount, partitionTypes, OptionalInt.empty());
+        }
+        else {
+            return createPartitioningHandleWithHiveType(partitionCount, ImmutableList.of(), OptionalInt.empty());
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2247]: Why do we have to create Hive compatible partitioning if partitionTypes is empty?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 7]
```diff
@@ -2142,9 +2165,9 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
             return Optional.empty();
         }
 
-        return Optional.of(new HivePartitioningHandle(
+        return Optional.of(createPartitioningHandleWithHiveType(
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2168]: Let's keep it compatible with native partitioning handle as well. 
Please check if the partitioning handle type is the same and if the column types are the same as well.  Also it looks like it makes sense to keep the constructor public, so it is easier to change the `maxCompatibleBucketCount` without performing extra logic.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 8]
```diff
@@ -2188,7 +2211,7 @@ public ConnectorTableLayoutHandle getAlternativeLayoutHandle(ConnectorSession se
         HiveBucketHandle bucketHandle = hiveLayoutHandle.getBucketHandle().get();
         ImmutableList<HiveType> bucketTypes = bucketHandle.getColumns().stream().map(HiveColumnHandle::getHiveType).collect(toImmutableList());
         checkArgument(
-                hivePartitioningHandle.getHiveTypes().equals(bucketTypes),
+                hivePartitioningHandle.getHiveTypes().get().equals(bucketTypes),
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2214]: Same comments as for `getCommonPartitioningHandle`.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 9]
```diff
@@ -2355,17 +2376,33 @@ private static Domain buildColumnDomain(ColumnHandle column, List<HivePartition>
...
         }
 
+        HivePartitioningHandle hivePartitioningHandle;
         List<String> bucketedBy = bucketProperty.get().getBucketedBy();
-        Map<String, HiveType> hiveTypeMap = tableMetadata.getColumns().stream()
-                .collect(toMap(ColumnMetadata::getName, column -> toHiveType(typeTranslator, column.getType())));
-        return Optional.of(new ConnectorNewTableLayout(
-                new HivePartitioningHandle(
+
+        HiveBucketProperty.BucketingType bucketingType = bucketProperty.get().getBucketingType();
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2382]: We never directly create temporary table. We always insert. Why the change to the `getNewTableLayout` is needed?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveNodePartitioningProvider.java [hunk 10]
```diff
@@ -49,8 +51,15 @@ public BucketFunction getBucketFunction(
...
-        List<HiveType> hiveTypes = handle.getHiveTypes();
-        return new HiveBucketFunction(bucketCount, hiveTypes);
+        HiveBucketProperty.BucketingType bucketingtype = handle.getBucketingType();
+        switch (bucketingtype) {
+            case PRESTOTYPE:
+                return createPrestoNativeBucketFunction(bucketCount, handle.getTypes().get());
+            case HIVETYPE:
+                return createHiveCompatibleBucketFunction(bucketCount, handle.getHiveTypes().get());
+            default:
+                throw new IllegalStateException(format("Unsupported bucketing type %s", bucketingtype));
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveNodePartitioningProvider.java:61]: `IllegalArgumentException`

Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveNodePartitioningProvider.java:61]: `"Unsupported bucketing type " + bucketingtype`

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java [hunk 11]
```diff
@@ -34,16 +36,42 @@
...
             @JsonProperty("bucketCount") int bucketCount,
-            @JsonProperty("sortedBy") List<SortingColumn> sortedBy)
+            @JsonProperty("sortedBy") List<SortingColumn> sortedBy,
+            @JsonProperty("types") Optional<List<Type>> types)
     {
         this.bucketedBy = ImmutableList.copyOf(requireNonNull(bucketedBy, "bucketedBy is null"));
         this.bucketCount = bucketCount;
         this.sortedBy = ImmutableList.copyOf(requireNonNull(sortedBy, "sortedBy is null"));
+        this.types = requireNonNull(types, "type is null");
+        if (types.isPresent()) {
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java:61]: Let's set it explicitly as a constructor parameter. Lets check if the `types` is set if the `bucketingType` is prestonative

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java [hunk 12]
```diff
@@ -34,16 +36,42 @@
...
+            bucketingType = BucketingType.PRESTOTYPE;
+            checkArgument(types.get().size() > 0, "types can not be empty if present");
+            checkArgument(types.get().size() == bucketedBy.size(), "the sizes of bucketedBy and types should match");
+        }
+        else {
+            bucketingType = BucketingType.HIVETYPE;
+        }
+    }
+
+    public enum BucketingType
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java:71]: Let's extract it as a standalone class. Let's call it `BucketFunctionType`

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java [hunk 13]
```diff
@@ -34,16 +36,42 @@
...
+    {
+        HIVETYPE,
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java:73]: `HIVE_COMPATIBLE`, `PRESTO_NATIVE`

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java [hunk 14]
```diff
@@ -34,16 +36,42 @@
     private final List<String> bucketedBy;
     private final int bucketCount;
     private final List<SortingColumn> sortedBy;
+    private final Optional<List<Type>> types;
+    private final BucketingType bucketingType;
+
+    public HiveBucketProperty(
```
Comment (arhimondr) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java:42]: How strongly do we feel about keeping this constructor? How many places have to be refactored if we are to remove it?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketFunction.java [hunk 15]
```diff
@@ -14,33 +14,71 @@
...
+    }
+
+    public static HiveBucketFunction createPrestoNativeBucketFunction(
+            int bucketCount,
+            List<Type> types)
+    {
+        return new HiveBucketFunction(bucketCount, Optional.empty(), Optional.of(types));
+    }
+
+    private HiveBucketFunction(
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketFunction.java:55]: Same comments as for other classes. Let's pass `bucketingType` as a parameter, and let's validate if the `hiveTypes` and `types` are set accordingly.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 16]
```diff
@@ -859,7 +861,24 @@ public ConnectorTableHandle createTemporaryTable(ConnectorSession session, List<
                         Sets.difference(ImmutableSet.copyOf(partitioning.getPartitionColumns()), allColumns)));
             }
             HivePartitioningHandle partitioningHandle = (HivePartitioningHandle) partitioning.getPartitioningHandle();
-            return new HiveBucketProperty(partitioning.getPartitionColumns(), partitioningHandle.getBucketCount(), ImmutableList.of());
+            List<String> partitionColumns = partitioning.getPartitionColumns();
+            if (partitionColumns.isEmpty() || !partitioningHandle.getTypes().isPresent()) {
+                return new HiveBucketProperty(
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:866]: Why do we need to use hive compatible bucket property if `partitionColumns.isEmpty() || !partitioningHandle.getTypes().isPresent()`?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 7 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2168]: Good point, I forget to check column types consistency. This method is marked as deprecated but I think the code is still in use, right?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 6 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2247]: It doesn't matter which one to use,  let me use Presto Native to make code simple.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 9 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2382]: Then this is not needed, will remove.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveNodePartitioningProvider.java (hunk 10 above)
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveNodePartitioningProvider.java:61]: `IllegalStateException` makes more sense:  `HivePartitioningHandle` has an invalid `BucketFunctionType`

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java (hunk 13 above)
Comment (viczhang861) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java:73]: How about `HIVE_TYPE` and `PRESTO_NATIVE`?  `COMP
ATIBLE` is a little confused , as there is no translation here.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java (hunk 14 above)
Comment (viczhang861) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java:42]: More than 10 call sites,   since "Optional<List<Type>> types" is optional, keeping the old constructor has some benefits.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 17]
```diff
@@ -2297,12 +2339,12 @@ private static Domain buildColumnDomain(ColumnHandle column, List<HivePartition>
             throw new PrestoException(NOT_SUPPORTED, "Writing to bucketed sorted Hive tables is disabled");
         }
 
-        HivePartitioningHandle partitioningHandle = new HivePartitioningHandle(
+        HivePartitioningHandle partitioningHandle = createHivePartitioningHandle(
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2342]: There are two places which create `HivePartitioningHandle` using `HiveBucketHandle and HiveBucketProperty`, a function is created to remove duplicated code.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 18]
```diff
@@ -2187,10 +2231,15 @@ public ConnectorTableLayoutHandle getAlternativeLayoutHandle(ConnectorSession se
         checkArgument(hiveLayoutHandle.getBucketHandle().isPresent(), "Hive connector only provides alternative layout for bucketed table");
         HiveBucketHandle bucketHandle = hiveLayoutHandle.getBucketHandle().get();
         ImmutableList<HiveType> bucketTypes = bucketHandle.getColumns().stream().map(HiveColumnHandle::getHiveType).collect(toImmutableList());
+        Optional<List<HiveType>> hiveTypes = hivePartitioningHandle.getHiveTypes();
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2234]: This can only be `HIVE_COMPATIBLE`, see explanation `checkArgument(
                hivePartitioningHandle.getBucketFunctionType().equals(HIVE_COMPATIBLE) &&
                        hiveTypes.isPresent(),
                "partitionColumns in HiveTableLayoutHandle must use a hive compatible bucket function");`

[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2234]: As per discussion offline we decided to keep an assertion, as this method shouldn't be used for temporary tables

**Comment by viczhang861** [comment]
@arhimondr 

- `adding an assertion to HiveMetadata#getTableMetadata to make sure the bucket function type is always HIVE_COMPATIBLE` --   this doesn't matter, temporary table also calls this function.

-  I have checked all call sites for `HiveBucketing#getHiveBuckets` and `HiveBucketing#getHiveBucket`
1.  For `HiveBucketFilter` - addressed
2. For `BucketAdapter` -  temporary table doesn't use this since bucket number is the same

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java (hunk 14 above)
Comment (viczhang861) [reply presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java]: Done

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 19]
```diff
@@ -287,7 +305,8 @@ private static int hashBytes(int initialValue, Slice bytes)
 
     public static Optional<HiveBucketFilter> getHiveBucketFilter(Table table, TupleDomain<ColumnHandle> effectivePredicate)
     {
-        if (!table.getStorage().getBucketProperty().isPresent()) {
+        Optional<HiveBucketProperty> hiveBucketProperty = table.getStorage().getBucketProperty();
+        if (!hiveBucketProperty.isPresent() || !hiveBucketProperty.get().getBucketFunctionType().equals(HIVE_COMPATIBLE)) {
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java:309]: This one is needed to return empty HiveBucketFilter:   ignore bucket pruning

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 20]
```diff
@@ -2559,6 +2603,27 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
         return toCompletableFuture(stagingFileCommitter.commitFiles(session, handle.getSchemaName(), handle.getTableName(), getPartitionUpdates(fragments)));
     }
 
+    private HivePartitioningHandle createHivePartitioningHandle(
+            int bucketCount,
+            HiveBucketHandle hiveBucketHandle,
+            Optional<HiveBucketProperty> bucketProperty,
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2609]: I initially use non-optional, but there was failed test due to empty `HiveBucketProperty`

[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2609]: In what case the `HiveBucketProperty` is missing?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java:309]: I'm not sure if it is safe to ignore it. There shouldn't be any bucket pruning for temporary tables though.

[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java:309]: Since code path exist for bucketed temporary table : `Optional<HiveBucketFilter> bucketFilter = shouldIgnoreTableBucketing ? Optional.empty() : getHiveBucketFilter(table, effectivePredicate);`    Thus we cannot use assertion but should ignore.

[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java:309]: Or add the check in call site: `shouldIgnoreTableBucketing || bucket function type == PRESTO_NATIVE ? Optional.empty() : getHiveBucketFilter(table, effectivePredicate);`
Same thing, let's keep it inside.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 21]
```diff
@@ -2142,19 +2174,31 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
             return Optional.empty();
         }
 
-        return Optional.of(new HivePartitioningHandle(
-                smallerBucketCount,
-                leftHandle.getHiveTypes(),
-                maxCompatibleBucketCount));
+        BucketFunctionType bucketFunctionType = leftHandle.getBucketFunctionType();
+        switch (bucketFunctionType) {
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2178]: Ok, I will remove the restriction to not use `HivePartitioningHandle` constructor directly and and more checking in constructor.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 22]
```diff
@@ -2559,6 +2603,27 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
         return toCompletableFuture(stagingFileCommitter.commitFiles(session, handle.getSchemaName(), handle.getTableName(), getPartitionUpdates(fragments)));
     }
 
+    private HivePartitioningHandle createHivePartitioningHandle(
+            int bucketCount,
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2607]: One place is using `getReadBucketCount()`, the other place is using `getTotalBucketCount()`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2609]: Here are some examples: https://api.travis-ci.com/v3/job/347805400/log.txt,   `HiveBucketProperty` is optional inside `Storage` class

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 23]
```diff
@@ -2559,6 +2603,27 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
...
+            HiveBucketHandle hiveBucketHandle,
+            Optional<HiveBucketProperty> bucketProperty,
+            OptionalInt maxCompatibleBucketCount)
+    {
+        if (bucketProperty.isPresent() &&
+                bucketProperty.get().getBucketFunctionType().equals(PRESTO_NATIVE)) {
```
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2613]: Since bucketProperty could be empty,  this function is a little different from other places:  if hive bucker property exists and indicates to use `PRESTO_NATIVE`, then use it, otherwise,  use the old way.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2602]: Okay, now I understand. The `bucketProperty` will be missing only if the table is "virtually bucketed". Let's have a special case for that to make it explicit, otherwise it is very confusing. 

Honestly i don't really like this method. It is very confusing. It is very hard to understand in what cases  `bucketProperty` is present, why `bucketCount`, `hiveBucketHandle` and `maxCompatibleBucketCount` may have different bucket count, etc. I would recommend to inline this method. It will result in a couple of lines of duplicated code, but it should make things much clearer.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2606]: It feels like the bucketProperty should be empty only if the table is virtually bucketed. Let's have a special case for that.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java:309]: It feels like this shortcut is not needed. Temporary tables are never filtered. Thus it seems like the method should return at `if (!bindings.isPresent()) {`, before reaching the `getHiveBuckets(table, bindings.get())` that would fail if the hash function is not `HIVE_COMPATIBLE`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 24]
```diff
@@ -324,7 +343,10 @@ private static int hashBytes(int initialValue, Slice bytes)
 
     private static Optional<Set<Integer>> getHiveBuckets(Table table, Map<ColumnHandle, Set<NullableValue>> bindings)
     {
-        List<String> bucketColumns = table.getStorage().getBucketProperty().get().getBucketedBy();
+        HiveBucketProperty hiveBucketProperty = table.getStorage().getBucketProperty().get();
+        checkArgument(hiveBucketProperty.getBucketFunctionType().equals(HIVE_COMPATIBLE),
+                "Expect hive compatible bucket function type to calculate bucket number based on Hive TypeInfo");
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java:348]: nit: `Expected bucketFunctionType to be HIVE_COMPATIBLE, got %s` ...

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java:309]: `bindings` is a non-empty Optional object for temporary table, query will fail if I revert this change. The shortcut has its value to avoid unneeded code path.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (viczhang861) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: The constructor is inline now.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 25]
```diff
@@ -2074,16 +2100,48 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
...
+                partitioningHandle = createHiveCompatiblePartitioningHandle(
+                        bucketCount,
+                        hiveBucketHandle.getColumns().stream()
+                                .map(HiveColumnHandle::getHiveType)
+                                .collect(toImmutableList()),
+                        maxCompatibleBucketCount);
+            }
+            else {
+                HiveBucketProperty bucketProperty = table.getStorage().getBucketProperty()
+                        .orElseThrow(() -> new NoSuchElementException("Bucket property should be set"));
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:2119]: nit: `new IllegalArgumentException("bucketProperty is expected to be present")`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java:309]: Okay, I just checked the code, and it seems to be safe to skip bucket filter.

Let's at least extract this condition into a separate branch and add a comment, e.g.:

```
if (!hiveBucketProperty.isPresent()) {
  return Optional.empty();
}
if(!hiveBucketProperty.get().getBucketFunctionType().equals(HIVE_COMPATIBLE)){
  // bucket filtering is only supported for tables bucketed with HIVE_COMPATIBLE hash function
  return Optional.empty();
}
```
